### PR TITLE
Pass connection context to transaction execution

### PR DESCRIPTION
**Pull request by matthiasbeyer**
My approach on/Targets #563.

### CODE REVIEW

File: src/action/manager.c [hunk 1]
```diff
@@ -159,8 +161,18 @@ ws_action_manager_process(
         }
 
         if (flags & WS_TRANSACTION_FLAGS_EXEC) {
+            struct ws_value_object_id* ctx = NULL;
+            if (opt_ctx) {
+                ctx = calloc(1, sizeof(*ctx));
+                if (!ctx) {
+                    //!< @todo ... aaahm... fuck! What to do?
```
Comment (neithernut) [inline src/action/manager.c]: Leave it alone, it should be initialized with `nil`.

File: src/action/manager.c [hunk 2]
```diff
@@ -159,8 +161,17 @@ ws_action_manager_process(
...
 
         if (flags & WS_TRANSACTION_FLAGS_EXEC) {
+            struct ws_value_object_id* ctx = NULL;
+            if (opt_ctx) {
+                ctx = calloc(1, sizeof(*ctx));
+                if (ctx) {
+                    ws_value_object_id_init(ctx);
+                    ws_value_object_id_set(ctx, opt_ctx);
+                } // else ... init with NULL...
+            }
```
Comment (neithernut) [inline src/action/manager.c]: Why are you allocating memory for the object id on the _heap_?

File: src/action/manager.c [hunk 3]
```diff
@@ -209,10 +220,18 @@ ws_action_manager_process(
             goto cleanup_name;
         }
 
+        struct ws_value_object_id* ctx = NULL;
+        if (opt_ctx) {
+            ctx = calloc(1, sizeof(*ctx));
+            ws_value_object_id_init(ctx);
+            ws_value_object_id_set(ctx, opt_ctx);
+        }
+
```
Comment (neithernut) [inline src/action/manager.c]: Same as above.

File: src/action/manager.c (hunk 2 above)
Comment (matthiasbeyer) [inline src/action/manager.c]: That's indeed a point.

File: src/action/manager.c [hunk 4]
```diff
@@ -159,8 +161,13 @@ ws_action_manager_process(
...
 
         if (flags & WS_TRANSACTION_FLAGS_EXEC) {
+            struct ws_value_object_id ctx;
+            ws_value_object_id_init(&ctx);
+            if (opt_ctx) {
+                ws_value_object_id_set(&ctx, opt_ctx);
+            }
             // execute the transaction
-            return run_transaction(transaction, NULL);
+            return run_transaction(transaction, NULL, &ctx.val);
```
Comment (neithernut) [inline src/action/manager.c]: You should pass either `NULL` or a nil value here if `opt_ctx==NULL`. The first option would probably be easier to implement.

File: src/action/manager.c [hunk 5]
```diff
@@ -209,10 +216,17 @@ ws_action_manager_process(
...
+        if (opt_ctx) {
+            ws_value_object_id_set(&ctx, opt_ctx);
+        }
+
         // now, finally, run the transaction
         struct ws_reply* reply;
         reply = run_transaction(transaction,
-                                &ws_event_get_context(event)->value);
+                                &ws_event_get_context(event)->value,
+                                &ctx.val);
```
Comment (neithernut) [inline src/action/manager.c]: Same here.

File: src/action/manager.c [hunk 6]
```diff
@@ -159,8 +161,15 @@ ws_action_manager_process(
         }
 
         if (flags & WS_TRANSACTION_FLAGS_EXEC) {
+            struct ws_value* vctx = NULL;
+            struct ws_value_object_id ctx;
+            ws_value_object_id_init(&ctx);
+            if (opt_ctx) {
+                ws_value_object_id_set(&ctx, opt_ctx);
+                vctx = &ctx.val;
+            }
```
Comment (neithernut) [inline src/action/manager.c]: You _do_ have this seven lines twice in your code. Maybe you should move them to the beginning of the function (e.g. always initialize `vcxt` and use it if needed).

Comment (matthiasbeyer) [inline src/action/manager.c]: I know. I think it would be more sufficient to keep them but split this function, as it is already over 100 loc. (without the doubled code nearly 100loc) and it is literally a `if() { /* do much stuff */ } else { /* do other stuff */ }`, which could be easily splitted...

Comment (neithernut) [inline src/action/manager.c]: You can still split the function _and_ have the code eduplicated:
The interface should not be changed (e.g. thee'll always be an `ws_action_manager_process()`), so I'd pull out the code into two functions (e.g. one for transactions, one for events). Those can _still_ get a pointer to the value as an parameter.
The split might be done in another PR.

Comment (matthiasbeyer) [inline src/action/manager.c]: So I will clean this up and then submit a follow-up PR to split the function.

**Comment by neithernut** [comment]
ACK for squashing.

**Comment by neithernut** [comment]
ACK, but please trigger the CI once more before merging (the last test is failing again).
